README.md
# ti_am623x_pru-icss

pru-icss-6.1.0/Release_Notes.txt
==========================================
PRU SOFTWARE SUPPORT PACKAGE RELEASE NOTES
==========================================

The PRU Software Support Package (PSSP) is not updated on a set schedule.
However, when TI updates the PSSP, we group commits by tag. Specific tags of
the PSSP are available for download here:
https://git.ti.com/cgit/pru-software-support-package/pru-software-support-package/

Release Notes for each tag are below. For additional details, reference the
individual commit messages.
e.g.,
$ git log --oneline
copy the git hash of the commit you are interested in
$ git show gitHash

v6.1.0
------
Released 2022-04-07

This release adds AM62x support to the PSSP. AM62x PRU is compatible with Linux
kernel 5.10 and later.
 * include/am62x (header files)
 * examples/am62x (Halt, Direct_Connect, MAC_Multiply_Accum, RPMsg)
 * labs/Getting_Started_Labs & labs/Hands_on_Labs/lab_5

Other changes:
 * Remove the "Protect" section from PRU Constant Table entries. Affected the
     linker command files for ICSSG devices (AM64x, AM65x, J7). Also applies to
     ICSSM v2.3 devices (which start with AM62x & AM263x).
 * If the "Protect" PRU Constant Table entry is reserved, then the reserved
     section must be less than 0x200. Reduce the arbitrary reserved section
     lengths in the linker command files for AM64x, AM65x, J7.
 * examples/j7: in TX_PRU linker command files, change PRU0/1_CTRL to
     TX_PRU0/1_CTRL. TI does NOT currently support PRU development on J7.


v6.0.1
------
Released 2021-11-24

Bugfix on top of v6.0.0.
PRU_Direct_Connect0 for PRU_ICSSG devices was disabling
interrupts instead of clearing the status of interrupts. Fix AM64x & AM65x by
writing to ICSS_INTC_ENA_STATUS_REGx instead of ICSS_INTC_ENABLE_CLR_REGx.


v6.0.0
------
Released 2021-10-17

This release supports Linux kernel 5.10 RemoteProc driver:
 * Port include/pru_types to be Linux kernel 5.10 compatible. Breaks kernel 5.4
     compatibility. Use PSSP v5.9.0 or earlier for Linux kernel 5.4.
 * Port existing projects in examples/ and labs/ to be compatible with
     Linux kernel 5.10. Applies to AM335x, AM437x, AM57xx, AM65x, K2G, J7ES1.

Add AM64x support. AM64x PRU is compatible with Linux kernel 5.10 and later.
 * include/am64x (header files)
 * examples/am64x (Halt, Direct_Connect, MAC_Multiply_Accum, RPMsg)
 * labs/Getting_Started_Labs & labs/Hands_on_Labs/lab_5

J721E PRU development is not supported by TI at the time of this tag. J721E
projects were ported to Linux 5.10 to enable community members who want to
develop J721E PRU without TI support.

K2G PRU development is only supported by TI up through the latest K2G Processor
SDK release (SDK 6.3, Linux kernel 4.19 at the time of this tag). Thus, K2G was
removed from the labs in v6.0.0. Other K2G projects were ported to Linux 5.10 to
enable community members who want to develop K2G PRU without TI support.

Update Direct_Connect examples:
 * Show how to pass INTC configuration to the Linux 5.10 RemoteProc driver
 * Add additional comments for how to pass interrupts and data between PRU cores
 * Add ReadMe.txt with FAQs about testing the example in Linux and loading the
   example from CCS or a non-Linux core.
 * Update am335x, am437x, am572x, k2g projects
 * Add projects for am64x, am65x

Other changes:
 * examples/<processor>/ReadMe.txt: add descriptions of each example
 * all Makefiles: new comments, remove RPMsg library from non-RPMsg projects
 * am65x: Add TX_PRU_CTRL to cmd files
 * TX_PRU_Halt: fix Makefile include_path
 * examples: Remove PRU_ARMtoPRU_Interrupt & PRU_PRUtoARM_Interrupt projects
 * examples/am437x: remove patches folder
 * examples/am65x: fixed date in RTU_Halt
 * labs/Getting_Started_Labs: Remove compiler optimization to make compiled
   firmware disassembly easier to read


v5.9.0
------
Released 2021-09-07

This is the last PSSP release that supports the Linux kernel 5.4 version of the
RemoteProc driver.

Add Release_Notes.txt


v5.8.0
------
Released 2021-03-31

Updated the linker command files so that there is proper alignment for
processors with 64 bit ARM cores (AM65x and j7es, examples & labs).

Updated AM65x Hands_on_Labs cmd files to integrate previous cmd linker file
changes.

Define DCFLAGS in the makefiles that use it (am65x & j7es).


v5.7.0
------
Released 2020-05-01

AM65x examples: replace SUBSYS with more descriptive SLICE in makefiles &
command linker files

Updated AM65x code for AM65x Silicon Revision 2.0 (examples and labs)

Updated software manifest & code to remove GPL-2.0 license

Update header files:
 * Use standard integer types in pru_virtio_ring.h
 * Remove types.h (unused)

READMEs: add link to TRMs


v5.6.0
------
Released 2019-12-16

Updated manifest file from v5.0 --> v6.0 (numbering unrelated to verson number
of PRU Software Support Package)

Added Getting Started Labs
 * Getting Started Labs are for AM335x, AM437x, AM57xx, and AM65xx
 * update labs folder makefiles

Several improvements to examples
 * PRU_gpioToggle toggles all GPO signals by default
 * AM65x examples, add ReadMe file


v5.5.0
------
Released 2019-09-18

Add support for Linux 4.19 with:
vring da field updates: (impacts resource table)
 * Add definition of FW_RSC_ADDR_ANY
 * Update labs & examples to use FW_RSC_ADDR_ANY

add support for j721E by:
	adding a header file for J721E ICSSG INTC
	adding RPMsg examples for J721E PRU & RTU cores
	adding Halt examples for J721E PRU, RTU, & TX_PRU cores

Add clarifying comments to examples:
 * No longer supported: PRU_ARMtoPRU_Interrupt, PRU_PRUtoARM_Interrupt
 * PRU_Hardware_UART: clarify how loopback works
 * ReadMe.txt updated to add new examples & documentation


v5.4.0
------
Released 2018-11-27

examples/am335x/PRU_ADC_onChip did not work out of the box in tag v5.3.0.
Implemented bug fixes.


v5.3.0
------
Released 2018-09-10

Add compatibility w/ Linux Kernel 4.14:
update all resource tables in examples and labs

Expanded hands-on labs:

workspace/pru0/main.c
/*
 * Copyright (C) 2022 Texas Instruments Incorporated - http://www.ti.com/
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *	* Redistributions of source code must retain the above copyright
 *	  notice, this list of conditions and the following disclaimer.
 *
 *	* Redistributions in binary form must reproduce the above copyright
 *	  notice, this list of conditions and the following disclaimer in the
 *	  documentation and/or other materials provided with the
 *	  distribution.
 *
 *	* Neither the name of Texas Instruments Incorporated nor the names of
 *	  its contributors may be used to endorse or promote products derived
 *	  from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdint.h>
#include <stdio.h>
#include <pru_intc.h>
#include <rsc_types.h>
#include <pru_rpmsg.h>
#include "resource_table.h"
#include "intc_map_0.h"

volatile register uint32_t __R31;

/* Host-0 Interrupt sets bit 30 in register R31 */
#define HOST_INT			((uint32_t) 1 << 30)

/*
 * The PRU-ICSS system events used for RPMsg are defined in the Linux device tree
 * PRU0 uses system event 16 (To ARM) and 17 (From ARM)
 * PRU1 uses system event 18 (To ARM) and 19 (From ARM)
 */
#define TO_ARM_HOST			16
#define FROM_ARM_HOST			17

/*
 * Using the name 'rpmsg-pru' will probe the rpmsg_pru driver found
 * at linux-x.y.z/drivers/rpmsg/rpmsg_pru.c
 */
#define CHAN_NAME                       "rpmsg-pru"
#define CHAN_DESC                       "Channel 30"
#define CHAN_PORT                       30

/*
 * Used to make sure the Linux drivers are ready for RPMsg communication
 * Found at linux-x.y.z/include/uapi/linux/virtio_config.h
 */
#define VIRTIO_CONFIG_S_DRIVER_OK	4

uint8_t payload[RPMSG_MESSAGE_SIZE];


/*
 * main.c
 */
void main(void)
{
	struct pru_rpmsg_transport transport;
	uint16_t src, dst, len;
	volatile uint8_t *status;

	/* Clear the status of the PRU-ICSS system event that the ARM will use to 'kick' us */
	CT_INTC.STATUS_CLR_INDEX_REG_bit.STATUS_CLR_INDEX = FROM_ARM_HOST;

	/* Make sure the Linux drivers are ready for RPMsg communication */
	status = &resourceTable.rpmsg_vdev.status;
	while (!(*status & VIRTIO_CONFIG_S_DRIVER_OK));

	/* Initialize the RPMsg transport structure */
	pru_rpmsg_init(&transport, &resourceTable.rpmsg_vring0, &resourceTable.rpmsg_vring1, TO_ARM_HOST, FROM_ARM_HOST);

	/* Create the RPMsg channel between the PRU and ARM user space using the transport structure. */
	while (pru_rpmsg_channel(RPMSG_NS_CREATE, &transport, CHAN_NAME, CHAN_DESC, CHAN_PORT) != PRU_RPMSG_SUCCESS);
	while (1) {
		/* Check bit 30 of register R31 to see if the ARM has kicked us */
		if (__R31 & HOST_INT) {
			/* Clear the event status */
			CT_INTC.STATUS_CLR_INDEX_REG_bit.STATUS_CLR_INDEX = FROM_ARM_HOST;
			/* Receive all available messages, multiple messages can be sent per kick */
			while (pru_rpmsg_receive(&transport, &src, &dst, payload, &len) == PRU_RPMSG_SUCCESS) {
				/* Echo the message back to the same address from which we just received */
				pru_rpmsg_send(&transport, dst, src, payload, len);
			}
		}
	}
}


pru-icss-6.1.0/include/am62x/pru_ctrl.h
/*
 * Copyright (C) 2022 Texas Instruments Incorporated - http://www.ti.com/
 *
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *	* Redistributions of source code must retain the above copyright
 *	  notice, this list of conditions and the following disclaimer.
 *
 *	* Redistributions in binary form must reproduce the above copyright
 *	  notice, this list of conditions and the following disclaimer in the
 *	  documentation and/or other materials provided with the
 *	  distribution.
 *
 *	* Neither the name of Texas Instruments Incorporated nor the names of
 *	  its contributors may be used to endorse or promote products derived
 *	  from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef _PRU_CTRL_H_
#define _PRU_CTRL_H_

/* PRU CTRL register set */
typedef struct {

	/* PRU_CTRL_CONTROL register bit field */
	union {
		volatile uint32_t CONTROL;

		volatile struct {
			uint32_t SOFT_RST_N : 1; // 0
			uint32_t PDSP_ENABLE : 1; // 1
			uint32_t PDSP_SLEEPING : 1; // 2
			uint32_t COUNTER_ENABLE : 1; // 3
			uint32_t RESTART : 1; // 4
			uint32_t rsvd5 : 3; // 7:5
			uint32_t SINGLE_STEP : 1; // 8
			uint32_t rsvd9 : 5; // 13:9
			uint32_t BIG_ENDIAN : 1; // 14
			uint32_t PDSP_STATE : 1; // 15
			uint32_t PCOUNTER_RST_VAL : 16; // 31:16
		} CONTROL_bit;
	}; // 0x0

	/* PRU_CTRL_STATUS register bit field */
	union {
		volatile uint32_t STATUS;

		volatile struct {
			uint32_t PCOUNTER : 16; // 15:0
			uint32_t rsvd16 : 16; // 31:16
		} STATUS_bit;
	}; // 0x4

	/* PRU_CTRL_WAKEUP_ENABLE register bit field */
	union {
		volatile uint32_t WAKEUP_ENABLE;

		volatile struct {
			uint32_t BITWISE_ENABLES : 32; // 31:0
		} WAKEUP_ENABLE_bit;
	}; // 0x8

	/* PRU_CTRL_CYCLE_COUNT register bit field */
	union {
		volatile uint32_t CYCLE_COUNT;

		volatile struct {
			uint32_t CYCLECOUNT : 32; // 31:0
		} CYCLE_COUNT_bit;
	}; // 0xc

	/* PRU_CTRL_STALL_COUNT register bit field */
	union {
		volatile uint32_t STALL_COUNT;

		volatile struct {
			uint32_t STALLCOUNT : 32; // 31:0
		} STALL_COUNT_bit;
	}; // 0x10

	uint8_t rsvd14[12]; // 0x14 - 0x1f

	/* PRU_CTRL_CONSTANT_TABLE_BLOCK_INDEX_0 register bit field */
	union {
		volatile uint32_t CONSTANT_TABLE_BLOCK_INDEX_0;

		volatile struct {
			uint32_t C24_BLK_INDEX : 8; // 7:0
			uint32_t rsvd8 : 8; // 15:8
			uint32_t C25_BLK_INDEX : 8; // 23:16
			uint32_t rsvd24 : 8; // 31:24
		} CONSTANT_TABLE_BLOCK_INDEX_0_bit;
	}; // 0x20

	/* PRU_CTRL_CONSTANT_TABLE_BLOCK_INDEX_1 register bit field */
	union {
		volatile uint32_t CONSTANT_TABLE_BLOCK_INDEX_1;

		volatile struct {
			uint32_t C26_BLK_INDEX : 8; // 7:0
			uint32_t rsvd8 : 8; // 15:8
			uint32_t C27_BLK_INDEX : 8; // 23:16
			uint32_t rsvd24 : 8; // 31:24
		} CONSTANT_TABLE_BLOCK_INDEX_1_bit;
	}; // 0x24

	/* PRU_CTRL_CONSTANT_TABLE_PROG_PTR_0 register bit field */
	union {
		volatile uint32_t CONSTANT_TABLE_PROG_PTR_0;

		volatile struct {
			uint32_t C28_POINTER : 16; // 15:0
			uint32_t C29_POINTER : 16; // 31:16
		} CONSTANT_TABLE_PROG_PTR_0_bit;
	}; // 0x28

	/* PRU_CTRL_CONSTANT_TABLE_PROG_PTR_1 register bit field */
	union {
		volatile uint32_t CONSTANT_TABLE_PROG_PTR_1;

		volatile struct {
			uint32_t C30_POINTER : 16; // 15:0
			uint32_t C31_POINTER : 16; // 31:16
		} CONSTANT_TABLE_PROG_PTR_1_bit;
	}; // 0x2c

} ctrl;

volatile __far ctrl CT_CTRL __attribute__((cregister("PRU_CTRL", near), peripheral);

#endif /* _PRU_CTRL_H_ */


pru-icss-6.1.0/include/am62x/pru_ecap.h
/*
 * Copyright (C) 2022 Texas Instruments Incorporated - http://www.ti.com/
 *
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *	* Redistributions of source code must retain the above copyright
 *	  notice, this list of conditions and the following disclaimer.
 *
 *	* Redistributions in binary form must reproduce the above copyright
 *	  notice, this list of conditions and the following disclaimer in the
 *	  documentation and/or other materials provided with the
 *	  distribution.
 *
 *	* Neither the name of Texas Instruments Incorporated nor the names of
 *	  its contributors may be used to endorse or promote products derived
 *	  from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef _PRU_ECAP_H_
#define _PRU_ECAP_H_

/* PRU ECAP register set */
typedef struct {

	/* PRU_ECAP_TSCNT register bit field */
	union {
		volatile uint32_t TSCNT;

		volatile struct {
			uint32_t TSCNT : 32; // 31:0
		} TSCNT_bit;
	}; // 0x0

	/* PRU_ECAP_CNTPHS register bit field */
	union {
		volatile uint32_t CNTPHS;

		volatile struct {
			uint32_t CNTPHS : 32; // 31:0
		} CNTPHS_bit;
	}; // 0x4

	/* PRU_ECAP_CAP1 register bit field */
	union {
		volatile uint32_t CAP1;

		volatile struct {
			uint32_t CAP1 : 32; // 31:0
		} CAP1_bit;
	}; // 0x8

	/* PRU_ECAP_CAP2 register bit field */
	union {
		volatile uint32_t CAP2;

		volatile struct {
			uint32_t CAP2 : 32; // 31:0
		} CAP2_bit;
	}; // 0xc

	/* PRU_ECAP_CAP3 register bit field */
	union {
		volatile uint32_t CAP3;

		volatile struct {
			uint32_t CAP3 : 32; // 31:0
		} CAP3_bit;
	}; // 0x10

	/* PRU_ECAP_CAP4 register bit field */
	union {
		volatile uint32_t CAP4;

		volatile struct {
			uint32_t CAP4 : 32; // 31:0
		} CAP4_bit;
	}; // 0x14

	uint8_t rsvd18[16]; // 0x18 - 0x27

	/* PRU_ECAP_ECCTL2_ECCTL1 register bit field */
	union {
		volatile uint32_t ECCTL2_ECCTL1;

		volatile struct {
			uint32_t CAP1POL : 1; // 0
			uint32_t CTRRST1 : 1; // 1
			uint32_t CAP2POL : 1; // 2
			uint32_t CTRRST2 : 1; // 3
			uint32_t CAP3POL : 1; // 4
			uint32_t CTRRST3 : 1; // 5
			uint32_t CAP4POL : 1; // 6
			uint32_t CTRRST4 : 1; // 7
			uint32_t CAPLDEN : 1; // 8
			uint32_t EVTFLTPS : 5; // 13:9
			uint32_t SOFT : 1; // 14
			uint32_t FREE : 1; // 15
			uint32_t CONT_ONESHT : 1; // 16
			uint32_t STOPVALUE : 2; // 18:17
			uint32_t REARM_RESET : 1; // 19
			uint32_t TSCNTSTP : 1; // 20
			uint32_t SYNCI_EN : 1; // 21
			uint32_t SYNCO_SEL : 2; // 23:22
			uint32_t SWSYNC : 1; // 24
			uint32_t CAP_APWM : 1; // 25
			uint32_t APWMPOL : 1; // 26
			uint32_t FILTER : 5; // 31:27
		} ECCTL2_ECCTL1_bit;
	}; // 0x28

	/* PRU_ECAP_ECFLG_ECEINT register bit field */
	union {
		volatile uint32_t ECFLG_ECEINT;

		volatile struct {
			uint32_t EN_RESV0 : 1; // 0
			uint32_t EN_CEVT1 : 1; // 1
			uint32_t EN_CEVT2 : 1; // 2
			uint32_t EN_CEVT3 : 1; // 3
			uint32_t EN_CEVT4 : 1; // 4
			uint32_t EN_CNTOVF : 1; // 5
			uint32_t EN_PRDEQ : 1; // 6
			uint32_t EN_CMPEQ : 1; // 7
			uint32_t EN__RESV1 : 8; // 15:8
			uint32_t FLAG_INT : 1; // 16
			uint32_t FLAG_CEVT1 : 1; // 17
			uint32_t FLAG_CEVT2 : 1; // 18
			uint32_t FLAG_CEVT3 : 1; // 19
			uint32_t FLAG_CEVT4 : 1; // 20
			uint32_t FLAG_CNTOVF : 1; // 21
			uint32_t FLAG_PRDEQ : 1; // 22
			uint32_t FLAG_CMPEQ : 1; // 23
			uint32_t FLAG_RESV0 : 8; // 31:24
		} ECFLG_ECEINT_bit;
	}; // 0x2c

	/* PRU_ECAP_ECCLR register bit field */
	union {
		volatile uint32_t ECCLR;

		volatile struct {
			uint32_t INT : 1; // 0
			uint32_t CEVT1 : 1; // 1
			uint32_t CEVT2 : 1; // 2
			uint32_t CEVT3 : 1; // 3
			uint32_t CEVT4 : 1; // 4
			uint32_t CNTOVF : 1; // 5
			uint32_t PRDEQ : 1; // 6
			uint32_t CMPEQ : 1; // 7
			uint32_t _RESV0 : 8; // 15:8
			uint32_t rsvd16 : 16; // 31:16
		} ECCLR_bit;
	}; // 0x30

	/* PRU_ECAP_ECFRC register bit field */
	union {
		volatile uint32_t ECFRC;

		volatile struct {
			uint32_t _RESV0 : 1; // 0
			uint32_t CEVT1 : 1; // 1
			uint32_t CEVT2 : 1; // 2
			uint32_t CEVT3 : 1; // 3
			uint32_t CEVT4 : 1; // 4
			uint32_t CNTOVF : 1; // 5
			uint32_t PRDEQ : 1; // 6
			uint32_t CMPEQ : 1; // 7
			uint32_t _RESV1 : 8; // 15:8
			uint32_t rsvd16 : 16; // 31:16
		} ECFRC_bit;
	}; // 0x34

	uint8_t rsvd38[36]; // 0x38 - 0x5b

	/* PRU_ECAP_PID register bit field */
	union {
		volatile uint32_t PID;

		volatile struct {
			uint32_t REVID : 32; // 31:0
		} PID_bit;
	}; // 0x5c

} ecap;

volatile __far ecap CT_ECAP __attribute__((cregister("PRU_ECAP", near), peripheral);

#endif /* _PRU_ECAP_H_ */
